spec/java/util/ArrayList.lsl

```diff
@@ -0,0 +1,1205 @@
+       nextIndex,
+       previousIndex,
+
+       // write operations
+       next,
+       remove,
+       forEachRemaining,
+       previous,
+       set,
+       add
+    ]
+
+    //constructors
+
+
+    constructor ListItr(index: int)
+    {
+        // #problem
+        //How invoke super() ??
+        cursor = index;
+    }
+
+
+    //methods
+
+
+    fun hasPrevious (): boolean
+    {
+        result = cursor != 0;
+    }
+
+
+    fun nextIndex (): int
+    {
+        result = cursor;
+    }
+
+
+    fun previousIndex (): int
+    {
+        result = cursor - 1;
+    }
+
+
+    fun hasNext (): boolean
+    {
+        result = cursor != self.parent.length;
+    }
+
+
+    fun next (): Object
+    {
+        self.parent._checkForComodification(expectedModCount);
+        var i = cursor;
+
+        if (i >= parent.length)
+        {
+            action THROW_NEW("java.util.NoSuchElementException", []);
+        }
+
+        // #problem
+        //I don't know what to do with ConcurrentModificationException(); ?
+
+        cursor = i + 1;
+        lastRet = i;
+        result = action LIST_GET(self.parent.storage, lastRet);
+    }
+
+
+    fun previous (): Object
+    {
+        self.parent._checkForComodification(expectedModCount);
+        var i = cursor - 1;
+
+        if (i < 0)
+        {
+            action THROW_NEW("java.util.NoSuchElementException", []);
+        }
+
+        // #problem
+        //I don't know what to do with ConcurrentModificationException(); ?
+
+        cursor = i;
+        lastRet = i;
+        result = action LIST_GET(self.parent.storage, lastRet);
+    }
+
+
+    fun remove (): void
+    {
+        if (lastRet < 0)
+        {
+            action THROW_NEW("java.lang.IllegalStateException", []);
+        }
+
+        self.parent._checkForComodification(expectedModCount);
+
+        // #problem
+        //What i must to do with try-catch in this method ?
+
+        self.parent._deleteElement(lastRet);
+        cursor = lastRet;
+        lastRet = -1;
+        expectedModCount = self.parent.modCount;
+    }
+
+    fun set (e: Object): void
+    {
+        if (lastRet < 0)
+        {
+            action THROW_NEW("java.lang.IllegalStateException", []);
+        }
+
+        self.parent._checkForComodification(expectedModCount);
+
+        // #problem
+        //What i must to do with try-catch in this method ?
+
+        self.parent._setElement(lastRet, e);
+    }
+
+
+    fun add (e: Object): void
+    {
+        self.parent._checkForComodification(expectedModCount);
+
+        // #problem
+        //What i must to do with try-catch in this method ?
+
+        var i = cursor;
+        self.parent._addElement(self.parent.length, e);
+        cursor = i + 1;
+        lastRet = -1;
+        expectedModCount = self.parent.modCount;
+    }
+
+
+    fun forEachRemaining (action: Consumer): void
+    {
+        action NOT_IMPLEMENTED();
+    }
+
+}
+
+
+
+@final
+@implements("java.util.Spliterator")
+@WrapperMeta(
```

Comment:
This should not be a wrapper as it is just an internal/private class.

Reply:
Fixed